zebra-network/src/constants.rs

```diff
@@ -90,6 +90,13 @@ pub const REQUEST_TIMEOUT: Duration = Duration::from_secs(20);
 /// nodes, and on testnet.
 pub const HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(3);
 
+/// The maximum time difference for two address book changes to be considered concurrent.
+///
+/// This timeout should be less than the reconnection and keepalive/heartbeat timeouts,
+/// but more than the amount of time between connection events and address book updates,
+/// even under heavy load. (In tests, we have observed delays up to 500ms.)
+pub const CONCURRENT_ADDRESS_CHANGE_PERIOD: Duration = HANDSHAKE_TIMEOUT;
```

Comment:
I think anything between 2-5 seconds would be appropriate.

Is there any way for the timestamps to reflect the precise order that changes should be applied?

Reply:
> Is there any way for the timestamps to reflect the precise order that changes should be applied? Could we do that in the follow-up work for the rest of the issue?

I don't think there is, because changes are sent by different tasks that must run concurrently: the candidate set dialer, and multiple connection message receiver closures. So we can't ensure that each received message is tagged with a time in order without a significant redesign to the entire network stack.

Even if we did, some machines have limited time resolution, so we can't guarantee that two messages on the same connection will always have different times. Sometimes the times would be equal, so we'd have to have state ordering logic anyway. (There are also known bugs impacting monotonic times which make them go backwards or stay equal.)

Reply:
> So we can't ensure that each received message is tagged with a time in order without a significant redesign to the entire network stack.

Also, some state change events are triggered by the application, so they can happen independently and simultaneously with peer messages. In that case, if we fail a connection, we don't want a peer to be able to overwrite it with "responded".

Reply:
> Are there any risks with this approach while it's shorter than the keepalive/heartbeat timeouts? (Could you add references to those constants too.)

I can't imagine any. The important state transition to preserve is `Failed/Responded -> AttemptPending`, which only happens after the reconnection timeout.

Reply:
> (Could you add references to those constants too.)

Done in [3e3e32c](https://github.com/ZcashFoundation/zebra/pull/6717/commits/3e3e32cd6613322b5b64ec3c57d64c34b7e974a6), along with docs for more edge cases and an increase to the concurrency period.

Reply:
> So we can't ensure that each received message is tagged with a time in order without a significant redesign to the entire network stack.

> Also, some state change events are triggered by the application, so they can happen independently and simultaneously with peer messages. In that case, if we fail a connection, we don't want a peer to be able to overwrite it with "responded".

Got it, thank you.

> Even if we did, some machines have limited time resolution, so we can't guarantee that two messages on the same connection will always have different times.

I think sequential response ids could fix that if there weren't other issues.